src/windows/list.cpp

```diff
@@ -0,0 +1,687 @@
+// - GUID_DEVINTERFACE_CDROM { 53F56308-B6BF-11D0-94F2-00A0C91EFB8B }
+// - GUID_DEVINTERFACE_USB_HUB { F18A0E88-C30C-11D0-8815-00A0C906BED8 }
+// - GUID_DEVINTERFACE_FLOPPY { 53F56311-B6BF-11D0-94F2-00A0C91EFB8B }
+// - GUID_DEVINTERFACE_WRITEONCEDISK { 53F5630C-B6BF-11D0-94F2-00A0C91EFB8B }
+// - GUID_DEVINTERFACE_TAPE { 53F5630B-B6BF-11D0-94F2-00A0C91EFB8B }
+// - GUID_DEVINTERFACE_USB_DEVICE { A5DCBF10-6530-11D2-901F-00C04FB951ED }
+// - GUID_DEVINTERFACE_VOLUME { 53F5630D-B6BF-11D0-94F2-00A0C91EFB8B }
+// - GUID_DEVINTERFACE_STORAGEPORT { 2ACCFE60-C130-11D2-B082-00A0C91EFB8B }
+//
+const GUID GUID_DEVICE_INTERFACE_DISK = {
+0x53F56307L, 0xB6BF, 0x11D0, { 0x94, 0xF2, 0x00, 0xA0, 0xC9, 0x1E, 0xFB, 0x8B }
+};
+
+const GUID GUID_DEVICE_INTERFACE_CDROM = {
+0x53F56308L, 0xB6BF, 0x11D0, { 0x94, 0xF2, 0x00, 0xA0, 0xC9, 0x1E, 0xFB, 0x8B }
+};
+
+const GUID GUID_DEVICE_INTERFACE_USB_HUB = {
+0xF18A0E88L, 0xC30C, 0x11D0, { 0x88, 0x15, 0x00, 0xA0, 0xC9, 0x06, 0xBE, 0xD8 }
+};
+
+const GUID GUID_DEVICE_INTERFACE_FLOPPY = {
+0x53F56311L, 0xB6BF, 0x11D0, { 0x94, 0xF2, 0x00, 0xA0, 0xC9, 0x1E, 0xFB, 0x8B }
+};
+
+const GUID GUID_DEVICE_INTERFACE_WRITEONCEDISK = {
+0x53F5630CL, 0xB6BF, 0x11D0, { 0x94, 0xF2, 0x00, 0xA0, 0xC9, 0x1E, 0xFB, 0x8B }
+};
+
+const GUID GUID_DEVICE_INTERFACE_TAPE = {
+0x53F5630BL, 0xB6BF, 0x11D0, { 0x94, 0xF2, 0x00, 0xA0, 0xC9, 0x1E, 0xFB, 0x8B }
+};
+
+const GUID GUID_DEVICE_INTERFACE_USB_DEVICE = {
+0xA5DCBF10L, 0x6530, 0x11D2, { 0x90, 0x1F, 0x00, 0xC0, 0x4F, 0xB9, 0x51, 0xED }
+};
+
+const GUID GUID_DEVICE_INTERFACE_VOLUME = {
+0x53F5630DL, 0xB6BF, 0x11D0, { 0x94, 0xF2, 0x00, 0xA0, 0xC9, 0x1E, 0xFB, 0x8B }
+};
+
+const GUID GUID_DEVICE_INTERFACE_STORAGEPORT = {
+0x2ACCFE60L, 0xC130, 0x11D2, { 0xB0, 0x82, 0x00, 0xA0, 0xC9, 0x1E, 0xFB, 0x8B }
+};
+
+const std::vector<std::string> USB_STORAGE_DRIVERS {
+  "USBSTOR", "UASPSTOR", "VUSBSTOR",
+  "RTUSER", "CMIUCR", "EUCR",
+  "ETRONSTOR", "ASUSSTPT"
+};
+
+const std::vector<std::string> GENERIC_STORAGE_DRIVERS {
+  "SCSI", "SD", "PCISTOR",
+  "RTSOR", "JMCR", "JMCF", "RIMMPTSK", "RIMSPTSK", "RIXDPTSK",
+  "TI21SONY", "ESD7SK", "ESM7SK", "O2MD", "O2SD", "VIACR"
+};
+
+// List of known virtual disk hardware IDs
+const std::vector<std::string> VHD_HARDWARE_IDS {
+  "Arsenal_________Virtual_",
+  "KernSafeVirtual_________",
+  "Msft____Virtual_Disk____",
+  "VMware__VMware_Virtual_S"
+};
+
+char* WCharToUtf8(const wchar_t* wstr) {
+  char *str = NULL;
+  size_t size = WideCharToMultiByte(CP_UTF8, 0, wstr, -1, NULL, 0, NULL, NULL);
+
+  if (size <= 1) {
+    return NULL;
+  }
+
+  if ((str = reinterpret_cast<char*>(calloc(size, 1))) == NULL) {
+    return NULL;
+  }
+
+  size_t utf8Size = WideCharToMultiByte(
+    CP_UTF8, 0, wstr, -1, str, size, NULL, NULL);
+
+  if (utf8Size != size) {
+    free(str);
+    return NULL;
+  }
+
+  return str;
+}
+
+char* GetEnumeratorName(HDEVINFO hDeviceInfo, SP_DEVINFO_DATA deviceInfoData) {
+  DWORD size;
+  DWORD dataType;
+  char buffer[MAX_PATH];
```

Comment:
MAX_PATH is a windows.h constant that denotes the maximum size of a path represented within Windows. So, yes, MAX_PATH will be big enough. Bigger is basically not an option.

Regarding the mix of A & W methods; some places don't need WideChars, like drive letters. There's just no point, because the conversion would add unnecessary complexity, and drive letters only go so far (A-Z isn't something that'd require Unicode).

Reply:
> MAX_PATH is a windows.h constant that denotes the maximum size of a path represented within Windows. So, yes, MAX_PATH will be big enough. Bigger is basically not an option.

That'd obviously be true if we're dealing with filesystem paths, but we're not dealing with filesystem paths here. However I agree that it seems very unlikely that any of the strings we're retrieving here will ever even _approach_ 260 characters, so I won't argue the point.